RubberduckTests/Grammar/AnnotationTests.cs

```diff
@@ -1,235 +1,52 @@
-            Assert.AreEqual(AnnotationType.PredeclaredId, annotation.AnnotationType);
-        }
-
-        [Category("Grammar")]
-        [Category("Annotations")]
-        [Test]
-        public void ObsoleteAnnotation_TypeIsObsolete()
-        {
-            var annotation = new ObsoleteAnnotation(new QualifiedSelection(), null, new[] { "param" });
-            Assert.AreEqual(AnnotationType.Obsolete, annotation.AnnotationType);
-        }
-
-        [Category("Grammar")]
-        [Category("Annotations")]
-        [Test]
-        public void IgnoreAnnotation_CanBeAppliedMultipleTimes()
-        {
-            var annotation = new IgnoreAnnotation(new QualifiedSelection(), null, null);
-            Assert.True(annotation.AllowMultiple);
-        }
-
-        [Category("Grammar")]
-        [Category("Annotations")]
-        [Test]
-        public void IgnoreModuleAnnotation_CanBeAppliedMultipleTimes()
-        {
-            var annotation = new IgnoreModuleAnnotation(new QualifiedSelection(), null, null);
-            Assert.True(annotation.AllowMultiple);
+        [TestCase(typeof(DefaultMemberAnnotation), "DefaultMember")]
+        [TestCase(typeof(DescriptionAnnotation), "Description")]
+        [TestCase(typeof(EnumeratorMemberAnnotation), "Enumerator")]
+        [TestCase(typeof(ExcelHotKeyAnnotation), "ExcelHotkey")]
+        [TestCase(typeof(ExposedModuleAnnotation), "Exposed")]
+        [TestCase(typeof(FolderAnnotation), "Folder")]
+        [TestCase(typeof(IgnoreAnnotation), "Ignore")]
+        [TestCase(typeof(IgnoreModuleAnnotation), "IgnoreModule")]
+        [TestCase(typeof(IgnoreTestAnnotation), "IgnoreTest")]
+        [TestCase(typeof(InterfaceAnnotation), "Interface")]
+        [TestCase(typeof(MemberAttributeAnnotation), "MemberAttribute")]
+        [TestCase(typeof(ModuleAttributeAnnotation), "ModuleAttribute")]
+        [TestCase(typeof(ModuleCleanupAnnotation), "ModuleCleanup")]
+        [TestCase(typeof(ModuleDescriptionAnnotation), "ModuleDescription")]
+        [TestCase(typeof(ModuleInitializeAnnotation), "ModuleInitialize")]
+        [TestCase(typeof(NoIndentAnnotation), "NoIndent")]
+        [TestCase(typeof(NotRecognizedAnnotation), "NotRecognized")]
+        [TestCase(typeof(ObsoleteAnnotation), "Obsolete")]
+        [TestCase(typeof(PredeclaredIdAnnotation), "PredeclaredId")]
+        [TestCase(typeof(TestCleanupAnnotation), "TestCleanup")]
+        [TestCase(typeof(TestInitializeAnnotation), "TestInitialize")]
+        [TestCase(typeof(TestMethodAnnotation), "TestMethod")]
+        [TestCase(typeof(TestModuleAnnotation), "TestModule")]
+        [TestCase(typeof(VariableDescriptionAnnotation), "VariableDescription")]
+        public void AnnotationTypes_MatchExpectedAnnotationNames(Type annotationType, string expectedName)
+        {
+            IAnnotation annotation = (IAnnotation)Activator.CreateInstance(annotationType);
+            Assert.AreEqual(expectedName, annotation.Name);
```

Comment:
Here, we explicitly enumerate all the types. However, we would be able to add new types and forget to add a test case here. Why not set it up to use a mini-CW (I think we still have a mock IoC somewhere in the project) to discover all annotations then assert that all found types conform?

Reply:
How would you set up the expected values then?

Reply:
Seems to me we have 2 choices here. We can make it a convention, which already is the case by requiring that the type names be name + `Annotation`. The expected can be then `nameof(theType)` minus the `Annotation`. Or, we can just blindly assert that the name is non-empty, since the CW registration will collect all the types that derives from the `IAnnotation` and isn't abstract.

Reply:
that convention is violated for `ExcelHotKeyAnnotation`, `ExposedMemberAnnotation` and possibly others as well...

Reply:
In which case, you could just set up a 2nd test that uses the mock IoC and assert that none of the names are empty. That will catch any new annotation that isn't correctly implemented. However, it won't catch those with wrong names (though we can guard against duplicate names, too). Still doesn't change the fact that we have to manually this unit test for every new annotation added.

Why not just fix them up to be conventional?

Reply:
I do not think we should fully enforce the convention. The names cannot change as they determine the annotation as seen in user code. The type names also become less descriptive if changed, e.g. `ExposedAnnotation` vs `ExposedModuleAnnotation`.

That said, we could set up two tests, one testing the exceptions explicitly and one testing all others for the convention. This will not require the mock IoC. We already provide the known annotation via the new `MockParser.GetWellKnownAnnotations`.

Reply:
added two test cases, one for non-emptyness of all "well known" annotations, one for the distinctness of the names.